src/main/java/slimeknights/tconstruct/smeltery/tileentity/FaucetTileEntity.java

```diff
@@ -0,0 +1,231 @@
+    if (toDrain != null && toFill != null) {
+      // can we drain?
+      FluidStack drained = toDrain.drain(144, IFluidHandler.FluidAction.SIMULATE);
+      if (!drained.isEmpty()) {
+        // can we fill
+        int filled = toFill.fill(drained, IFluidHandler.FluidAction.SIMULATE);
+        if (filled > 0) {
+          // drain the liquid and transfer it, buffer the amount for delay
+          this.drained = toDrain.drain(filled, IFluidHandler.FluidAction.EXECUTE);
+          this.isPouring = true;
+          pour();
+
+          // sync to clients
+          if (!getWorld().isRemote()) { // TODO: getWorld() instanceof WorldServer
+            // TODO: TinkerNetwork.sendToClients((WorldServer) getWorld(), pos, new FaucetActivationPacket(pos, drained));
+          }
+
+          return;
+        }
+      }
+    }
+    // draining unsuccessful
+    reset();
+  }
+
+  protected void pour() {
+    if (drained.isEmpty()) {
+      return;
+    }
+
+    IFluidHandler toFill = getFluidHandler(pos.down(), Direction.UP);
+    if (toFill != null) {
+      FluidStack fillStack = drained.copy();
+      fillStack.setAmount(Math.min(drained.getAmount(), 6));
+
+      // can we fill?
+      int filled = toFill.fill(fillStack, IFluidHandler.FluidAction.SIMULATE);
+      if (filled > 0) {
+        // transfer it
+        this.drained.shrink(filled);
+        fillStack.setAmount(filled);
+        toFill.fill(fillStack, IFluidHandler.FluidAction.EXECUTE);
+      }
+    }
+    else {
+      // filling TE got lost. reset. all liquid buffered is lost.
+      reset();
+    }
+  }
+
+  protected void reset() {
+    isPouring = false;
+    stopPouring = false;
+    drained = null;
+    direction = Direction.DOWN;
+    lastRedstoneState = false;
+  }
+
+  protected IFluidHandler getFluidHandler(BlockPos pos, Direction direction) {
+    TileEntity te = getWorld().getTileEntity(pos);
+    if (te != null) {
+      return te.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, direction).orElse(null);
+    }
+    return null;
+  }
+
+  @Override
+  @OnlyIn(Dist.CLIENT)
+  public AxisAlignedBB getRenderBoundingBox() {
+    return new AxisAlignedBB(pos.getX(), pos.getY() - 1, pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1);
+  }
+
+  @Override
+  public CompoundNBT write(CompoundNBT compound) {
+    compound = super.write(compound);
+    if (drained != null) {
+      drained.writeToNBT(compound);
+      compound.putInt("direction", direction.getIndex());
+    }
+    return compound;
+  }
+
+  @Override
+  public void read(CompoundNBT compound) {
+    super.read(compound);
+    drained = FluidStack.loadFluidStackFromNBT(compound);
+
+    if (drained != null) {
+      isPouring = true;
+      direction = Direction.values()[compound.getInt("direction")];
+      stopPouring = compound.getBoolean("stop");
+    }
+    else {
+      reset();
+    }
+  }
+
+  public void onActivationPacket(FluidStack fluid) {
```

Comment:
I know we do not have a ton of java docs in 1.12, but ideally we should have them. My rule of thumb is each method should have either a javadoc or an `@Override`

Reply:
Considering to do something akin to FluidUpdatePacket, make it an interface
and document it there. Sound fine?

Reply:
If we use it in multiple places, make an interface. I doubt the faucet packet will be used again elsewhere.

Reply:
Think we're good there. If needed we can factor it out later.